### PR TITLE
Feature/queue

### PR DESCRIPTION
**Pull request by Toflar**

### CODE REVIEW

File: dca/tl_nc_gateway.php [hunk 1]
```diff
@@ -156,6 +160,58 @@
...
+        'queue_targetGateway' => array
+        (
+            'label'                   => &$GLOBALS['TL_LANG']['tl_nc_gateway']['queue_targetGateway'],
+            'exclude'                 => true,
+            'filter'                  => true,
+            'inputType'               => 'select',
+            'options_callback'        => function() {
+                $options = array();
+
+                $gateways = \Database::getInstance()->prepare('SELECT id,title FROM tl_nc_gateway WHERE type!=?')
```
Comment (aschempp) [inline dca/tl_nc_gateway.php]: Why not use the models here?

File: dca/tl_nc_notification.php [hunk 2]
```diff
@@ -112,7 +112,7 @@
     'palettes' => array
     (
         '__selector__'                => array('type'),
-        'default'                     => '{title_legend},title,type',
+        'default'                     => '{title_legend},title,type;',
```
Comment (aschempp) [inline dca/tl_nc_notification.php]: It is not common in Contao to add a `;` at the end of a palette. If anyone would append to existing palettes, it should always be done with `;{…legend}…`

File: dca/tl_nc_queue.php [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                'href'                => 'act=show',
+                'icon'                => 'show.gif'
+            )
+        )
+    ),
+
+    // Palettes
+    'palettes' => array
+    (
+        'default'                     => '{title_legend},title,type'
```
Comment (aschempp) [inline dca/tl_nc_queue.php]: You don't need palettes if the records are not editable. But I think the `palettes` array must be there (just keep it empty).

**Comment by aschempp** [comment]
Looks good so far. Only one important question:

Currently we store the tokens in the queue. That means the message is actually rendered upon submitting. This will result in potentially different content than rendering instantly. Think about using insert tags like `{{date::…}}` in the message. What can we do about that?

**Comment by Toflar** [comment]
I thought about that too but that's not a problem we can solve I guess. For example, you might be right with the InsertTag `{{date::…}}` in the case when somebody would write

> The products you ordered on {{date::…}}

In this case it would be correct to pre-render them before adding it to the queue. However, if you wrote something like

> This e-mail has been sent on {{date::…}}

it would be correct to post-render. So to be honest, I think the current approach - while it might cause some problems depending on how people use InsertTags and stuff - is the right one. Because in my example the right way would be to provide a token for `##order_date##` or so.

**Comment by aschempp** [comment]
Hmm, you might be right about that. I mean the queue is only useful for mass *mailing*, and in that case it would very likely make sense to use the information of when the email was sent and not when it was generated. Obviously, we don't know though :D

**Comment by Toflar** [comment]
Yes, and also the behaviour strongly depends on how the e-mails are being queued. I mean, let's assume you have an InsertTag providing the current month and you have 2'000 e-mails and your cron only sends 100 e-mails per day. Then maybe you start sending at the end of the month so the first 4 days will get the right InsertTag replacement, and the others won't. There are endless scenarios where an InsertTag might or might not work properly :) If we do it that way, then we can just say "well, if you want a certain date that should not be changed, you better provide it as a simple token". So I'd merge it into develop then before we start forgetting about it again?

**Comment by aschempp** [comment]
Fine with me, what about @qzminski?

Also, I assume you did test it in a real-world installation? ^^

File: dca/tl_nc_queue.php (hunk 3 above)
Comment (qzminski) [inline dca/tl_nc_queue.php]: According to me the array does not need to be there if the records won't be altered through a regular back end options.

File: dca/tl_nc_gateway.php (hunk 1 above)
Comment (qzminski) [inline dca/tl_nc_gateway.php]: Not really needed and more efficient.

**Comment by qzminski** [comment]
Looks good to me.

File: dca/tl_nc_gateway.php (hunk 1 above)
Comment (Toflar) [inline dca/tl_nc_gateway.php]: Yeppa.

File: dca/tl_nc_queue.php (hunk 3 above)
Comment (Toflar) [inline dca/tl_nc_queue.php]: Fixed.

**Comment by Toflar** [comment]
Merged in 3613a1db8c1ac89a835a38622f8ce35aaf4f0474.

**Pull request closed without merging**